File: tools/RLMResearchPlannerPWA/src/language-pack.js
export const LANGUAGE_PACK_DOCUMENT_TYPE = "RLMResearchPlanner.language-pack";
export const LANGUAGE_PACK_SCHEMA_VERSION = 1;
export const LANGUAGE_PACK_SECTIONS = [
  "messages", "categories", "research", "buildings", "effects", "effect_labels", "effect_values",
  "resources", "talents", "talent_effects", "talent_presets", "talent_preset_descriptions",
];
export const PROTECTED_MESSAGE_KEYS = new Set(["app.disclaimer"]);
const STORAGE_KEY = "rlm-research-planner-pwa.language-packs.v1";
const LOCALE_PATTERN = /^[A-Za-z]{2,8}(?:-[A-Za-z0-9]{1,8})*$/;
const RTL_LANGUAGES = new Set(["ar", "arc", "ckb", "dv", "fa", "he", "ks", "nqo", "ps", "sd", "syr", "ug", "ur", "yi"]);

export function normalizeLocale(value) {
  const locale = String(value || "").trim().replaceAll("_", "-");
  if (!LOCALE_PATTERN.test(locale)) throw new Error("locale must be a BCP 47 language tag such as fr-FR or ar");
  return locale.split("-").map((part, index) => {
    if (index === 0) return part.toLocaleLowerCase("en-US");
    if (part.length === 2 && /^[A-Za-z]+$/.test(part)) return part.toLocaleUpperCase("en-US");
    if (part.length === 4 && /^[A-Za-z]+$/.test(part)) return `${part[0].toLocaleUpperCase("en-US")}${part.slice(1).toLocaleLowerCase("en-US")}`;
    return part;
  }).join("-");
}

export function defaultDirection(locale) {
  return RTL_LANGUAGES.has(String(locale).split("-", 1)[0].toLocaleLowerCase("en-US")) ? "rtl" : "ltr";
}

export function selectPreferredLocale(preferredLocales, availableLocales, fallbackLocale = "en-US") {
  const available = [];
  for (const value of availableLocales || []) {
    try {
      const locale = normalizeLocale(value);
      if (!available.includes(locale)) available.push(locale);
    } catch { /* Ignore invalid installed locale identifiers. */ }
  }
  const fallback = normalizeLocale(fallbackLocale);
  if (!available.length) return fallback;

  const preferred = [];
  for (const value of preferredLocales || []) {
    try {
      const locale = normalizeLocale(value);
      if (!preferred.includes(locale)) preferred.push(locale);
    } catch { /* Ignore invalid browser language identifiers. */ }
  }

  for (const locale of preferred) {
    const exact = available.find((candidate) => candidate.toLocaleLowerCase("en-US") === locale.toLocaleLowerCase("en-US"));
    if (exact) return exact;
    const language = locale.split("-", 1)[0].toLocaleLowerCase("en-US");
    const baseMatch = available.find((candidate) => candidate.split("-", 1)[0].toLocaleLowerCase("en-US") === language);
    if (baseMatch) return baseMatch;
  }

  return available.find((candidate) => candidate.toLocaleLowerCase("en-US") === fallback.toLocaleLowerCase("en-US")) || available[0];
}

function translatedText(value, section, key, trusted = false) {
  const sourceText = String(value && typeof value === "object" ? value.source || "" : "");
  const text = String(value && typeof value === "object" ? value.text || "" : value || "").trim();
  if (!trusted && /[<>]/u.test(text)) throw new Error(`${section}.${key}: translated text must not contain HTML markup`);
  if ([...text].some((character) => character.charCodeAt(0) < 32 && !"\t\n\r".includes(character))) throw new Error(`${section}.${key}: translated text contains a control character`);
  if (text.length > 20000) throw new Error(`${section}.${key}: translated text is too long`);
  if (section === "messages" && text && sourceText) {
    const placeholders = (input) => [...input.matchAll(/\{([A-Za-z0-9_]+)\}/g)].map((match) => match[1]).sort();
    if (JSON.stringify(placeholders(text)) !== JSON.stringify(placeholders(sourceText))) {
      throw new Error(`${section}.${key}: translated text must keep the source placeholders`);
    }
  }
  return text;
}

export function languagePackFromPayload(raw, { trusted = false } = {}) {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) throw new Error("language pack must be a JSON object");
  if (raw.document_type !== LANGUAGE_PACK_DOCUMENT_TYPE) throw new Error("unsupported language pack document_type");
  if (Number(raw.schema_version) !== LANGUAGE_PACK_SCHEMA_VERSION) throw new Error("unsupported language pack schema_version");
  const locale = normalizeLocale(raw.locale);
  const direction = String(raw.direction || defaultDirection(locale)).trim().toLocaleLowerCase("en-US");
  if (!new Set(["ltr", "rtl"]).has(direction)) throw new Error("direction must be ltr or rtl");
  const name = String(raw.name || locale).trim();
  if (!name || name.length > 100) throw new Error("name is required and must be at most 100 characters");
  const sections = {};
  for (const section of LANGUAGE_PACK_SECTIONS) {
    const source = raw[section] || {};
    if (!source || typeof source !== "object" || Array.isArray(source)) throw new Error(`${section} must be an object`);
    sections[section] = Object.fromEntries(Object.entries(source).flatMap(([rawKey, value]) => {
      const key = String(rawKey).trim();
      if (!key || key.length > 300) throw new Error(`${section} contains an invalid key`);
      if (!trusted && section === "messages" && PROTECTED_MESSAGE_KEYS.has(key)) return [];
      const text = translatedText(value, section, key, trusted);
      return text ? [[key, text]] : [];
    }));
  }
  return {
    locale,
    name,
    direction,
    fallbackLocale: normalizeLocale(raw.fallback_locale || "en-US"),
    author: String(raw.author || "").trim().slice(0, 200),
    license: String(raw.license || "").trim().slice(0, 200),
    datasetId: String(raw.catalog_dataset_id || "").trim().slice(0, 200),
    effectSeparator: Object.hasOwn(raw, "effect_separator") ? String(raw.effect_separator ?? "").slice(0, 20) : null,
    sections,
    source: raw,
  };
}

export function loadLanguagePacks(storage = localStorage) {
  try {
    const values = JSON.parse(storage.getItem(STORAGE_KEY) || "[]");
    return Object.fromEntries((Array.isArray(values) ? values : []).flatMap((raw) => {
      try { const pack = languagePackFromPayload(raw); return [[pack.locale, pack]]; }
      catch { return []; }
    }));
  } catch { return {}; }
}

function saveLanguagePacks(packs, storage) {
  storage.setItem(STORAGE_KEY, JSON.stringify(Object.values(packs).map((pack) => pack.source)));
}

export function installLanguagePack(raw, storage = localStorage) {
  const pack = languagePackFromPayload(raw);
  const packs = loadLanguagePacks(storage);
  packs[pack.locale] = pack;
  saveLanguagePacks(packs, storage);
  return pack;
}

export function removeLanguagePack(locale, storage = localStorage) {
  const normalized = normalizeLocale(locale);
  const packs = loadLanguagePacks(storage);
  if (!packs[normalized]) return false;
  delete packs[normalized];
  saveLanguagePacks(packs, storage);
  return true;
}

export function packText(pack, section, key, fallback = "") {
  return String(pack?.sections?.[section]?.[key] || fallback);
}

const entry = (source) => ({ source: String(source || ""), text: "" });

export function languagePackTemplate({ catalog, castleCatalog, talentCatalog, messages, fallbackPack = null }) {
  const categories = Object.fromEntries(catalog.categories.map((category) => [category.id, entry(catalog.sourceCategoryTitle(category, "en-US"))]));
  const research = Object.fromEntries([...catalog.nodes.values()].map((node) => [node.id, entry(catalog.sourceNodeName(node, "en-US"))]));
  const effects = Object.fromEntries([...catalog.nodes.values()].filter((node) => node.effectLabel).map((node) => [node.id, entry(node.effectLabel)]));
  const buildings = Object.fromEntries(castleCatalog.order.map((buildingId) => [buildingId, entry(castleCatalog.sourceBuildingName(buildingId, "en-US"))]));
  for (const [key, value] of Object.entries(fallbackPack?.sections?.buildings || {})) {
    if (!Object.hasOwn(buildings, key)) buildings[key] = entry(value);
  }
  const resources = Object.fromEntries(Object.entries(messages).filter(([key]) => key.startsWith("resource.")).map(([key, value]) => [key.slice("resource.".length), entry(value)]));
  const talents = Object.fromEntries([...(talentCatalog?.talents?.values?.() || [])].map((talent) => [talent.id, entry(talentCatalog.talentName(talent, "en-US"))]));
  const talentEffects = Object.fromEntries([...(talentCatalog?.talents?.values?.() || [])].map((talent) => [talent.id, entry(talentCatalog.effectName(talent, "en-US"))]));
  const talentPresets = Object.fromEntries((talentCatalog?.presets || []).map((preset) => [preset.id, entry(talentCatalog.presetName(preset, "en-US"))]));
  const talentPresetDescriptions = Object.fromEntries((talentCatalog?.presets || []).map((preset) => [preset.id, entry(talentCatalog.presetDescription(preset, "en-US"))]));
  return {
    document_type: LANGUAGE_PACK_DOCUMENT_TYPE,
    schema_version: LANGUAGE_PACK_SCHEMA_VERSION,
    locale: "xx",
    name: "New language",
    direction: "ltr",
    fallback_locale: "en-US",
    author: "",
    license: "",
    catalog_dataset_id: catalog.datasetId || "",
    effect_separator: " ",
    messages: Object.fromEntries(Object.entries(messages).filter(([key]) => !PROTECTED_MESSAGE_KEYS.has(key)).sort(([a], [b]) => a.localeCompare(b)).map(([key, value]) => [key, entry(value)])),
    categories,
    research,
    buildings,
    effects,
    effect_labels: Object.fromEntries(Object.entries(fallbackPack?.sections?.effect_labels || {}).map(([key, value]) => [key, entry(value)])),
    effect_values: Object.fromEntries(Object.entries(fallbackPack?.sections?.effect_values || {}).map(([key, value]) => [key, entry(value)])),
    resources,
    talents,
    talent_effects: talentEffects,
    talent_presets: talentPresets,
    talent_preset_descriptions: talentPresetDescriptions,
  };
}

export function localeManifestFromPayload(raw) {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) throw new Error("locale manifest must be a JSON object");
  if (raw.document_type !== "RLMResearchPlanner.locale-manifest" || Number(raw.schema_version) !== 1) throw new Error("unsupported locale manifest");
  const seen = new Set();
  const locales = (Array.isArray(raw.locales) ? raw.locales : []).map((entry) => {
    const locale = normalizeLocale(entry?.locale);
    if (seen.has(locale)) throw new Error(`duplicate bundled locale: ${locale}`);
    seen.add(locale);
    const name = String(entry?.name || "").trim();
    const direction = String(entry?.direction || defaultDirection(locale)).trim().toLocaleLowerCase("en-US");
    const path = String(entry?.path || "").trim();
    if (!name || name.length > 100) throw new Error(`${locale}: locale name is invalid`);
    if (!new Set(["ltr", "rtl"]).has(direction)) throw new Error(`${locale}: direction must be ltr or rtl`);
    if (!path || /[\\/]/u.test(path) || path === "." || path === "..") throw new Error(`${locale}: locale path is invalid`);
    return { locale, name, direction, path };
  });
  if (!locales.length) throw new Error("locale manifest must register at least one locale");
  const fallbackLocale = normalizeLocale(raw.fallback_locale || "en-US");
  if (!seen.has(fallbackLocale)) throw new Error("locale manifest fallback is not registered");
  return { fallbackLocale, locales, byLocale: Object.fromEntries(locales.map((entry) => [entry.locale, entry])) };
}

export async function loadBundledLanguagePacks(manifestUrl, version = "") {
  const response = await fetch(manifestUrl, { cache: "no-store" });
  if (!response.ok) throw new Error(`language manifest could not be loaded (${response.status})`);
  const manifest = localeManifestFromPayload(await response.json());
  const baseUrl = new URL(manifestUrl, globalThis.location?.href || "http://localhost/");
  const entries = await Promise.all(manifest.locales.map(async (entry) => {
    const url = new URL(entry.path, baseUrl);
    if (version) url.searchParams.set("v", version);
    const localeResponse = await fetch(url, { cache: "no-store" });
    if (!localeResponse.ok) throw new Error(`${entry.locale}: language pack could not be loaded (${localeResponse.status})`);
    const pack = languagePackFromPayload(await localeResponse.json(), { trusted: true });
    if (pack.locale !== entry.locale || pack.name !== entry.name || pack.direction !== entry.direction) throw new Error(`${entry.locale}: language pack metadata does not match its manifest entry`);
    return [entry.locale, pack];
  }));
  return { manifest, packs: Object.fromEntries(entries) };
}

export function mergeLanguagePacks(...values) {
  const packs = values.filter(Boolean);
  if (!packs.length) return null;
  const selected = packs[packs.length - 1];
  const effectSeparator = [...packs].reverse().find((pack) => pack.effectSeparator !== null && pack.effectSeparator !== undefined)?.effectSeparator ?? " ";
  return {
    ...selected,
    effectSeparator,
    sections: Object.fromEntries(LANGUAGE_PACK_SECTIONS.map((section) => [
      section,
      Object.assign({}, ...packs.map((pack) => pack.sections?.[section] || {})),
    ])),
  };
}

export function resolveLanguagePack(locale, bundledPacks, customPacks, fallbackLocale = "en-US") {
  const bundled = bundledPacks || {};
  const custom = customPacks || {};
  const layers = [];
  const visiting = new Set();
  const addLocale = (candidate) => {
    if (!candidate || visiting.has(candidate)) return;
    visiting.add(candidate);
    const selected = custom[candidate] || bundled[candidate];
    if (selected?.fallbackLocale && selected.fallbackLocale !== candidate) addLocale(selected.fallbackLocale);
    if (bundled[candidate]) layers.push(bundled[candidate]);
    if (custom[candidate]) layers.push(custom[candidate]);
  };
  addLocale(fallbackLocale);
  addLocale(locale);
  return mergeLanguagePacks(...layers);
}

export function applyDocumentLanguage(locale, direction = defaultDirection(locale)) {
  document.documentElement.lang = locale;
  document.documentElement.dir = direction;
  document.body?.setAttribute("dir", direction);
}

export function translateStatic(root, messages) {
  root.querySelectorAll("[data-i18n]").forEach((element) => {
    const value = messages[element.dataset.i18n];
    if (value) element.textContent = value;
  });
  root.querySelectorAll("[data-i18n-placeholder]").forEach((element) => {
    const value = messages[element.dataset.i18nPlaceholder];
    if (value) element.setAttribute("placeholder", value);
  });
  root.querySelectorAll("[data-i18n-title]").forEach((element) => {
    const value = messages[element.dataset.i18nTitle];
    if (value) element.setAttribute("title", value);
  });
  root.querySelectorAll("[data-i18n-aria]").forEach((element) => {
    const value = messages[element.dataset.i18nAria];
    if (value) element.setAttribute("aria-label", value);
  });
}

File: src/rlm_research_planner/services/language_pack.py
from __future__ import annotations

import json
import re
from string import Formatter
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, Mapping

from rlm_research_planner.domain.models import MasterData
from rlm_research_planner.domain.observations import ResearchTreeObservation
from rlm_research_planner.services.castle_planning import CastleCatalog


LANGUAGE_PACK_DOCUMENT_TYPE = "RLMResearchPlanner.language-pack"
LANGUAGE_PACK_SCHEMA_VERSION = 1
LANGUAGE_PACK_SECTIONS = (
    "messages",
    "categories",
    "research",
    "buildings",
    "effects",
    "effect_labels",
    "effect_values",
    "resources",
    "talents",
    "talent_effects",
    "talent_presets",
    "talent_preset_descriptions",
)
PROTECTED_MESSAGE_KEYS = frozenset(("app.disclaimer",))
_LOCALE_PATTERN = re.compile(r"^[A-Za-z]{2,8}(?:-[A-Za-z0-9]{1,8})*$")
_RTL_LANGUAGES = frozenset(
    ("ar", "arc", "ckb", "dv", "fa", "he", "ks", "nqo", "ps", "sd", "syr", "ug", "ur", "yi")
)


class LanguagePackError(ValueError):
    pass


@dataclass(frozen=True)
class LanguagePack:
    locale: str
    name: str
    direction: str
    fallback_locale: str
    author: str
    license_name: str
    dataset_id: str
    effect_separator: str | None
    sections: Mapping[str, Mapping[str, str]]

    def text(self, section: str, key: str) -> str:
        return str(self.sections.get(section, {}).get(key, "")).strip()

    def to_dict(self) -> dict[str, object]:
        return {
            "document_type": LANGUAGE_PACK_DOCUMENT_TYPE,
            "schema_version": LANGUAGE_PACK_SCHEMA_VERSION,
            "locale": self.locale,
            "name": self.name,
            "direction": self.direction,
            "fallback_locale": self.fallback_locale,
            "author": self.author,
            "license": self.license_name,
            "catalog_dataset_id": self.dataset_id,
            **(
                {"effect_separator": self.effect_separator}
                if self.effect_separator is not None
                else {}
            ),
            **{
                section: {
                    key: {"source": "", "text": value}
                    for key, value in sorted(self.sections.get(section, {}).items())
                }
                for section in LANGUAGE_PACK_SECTIONS
            },
        }


@dataclass(frozen=True)
class BundledLocale:
    locale: str
    name: str
    direction: str
    path: Path
    pack: LanguagePack


@dataclass(frozen=True)
class BundledLocaleManifest:
    fallback_locale: str
    locales: tuple[BundledLocale, ...]

    @property
    def by_locale(self) -> dict[str, BundledLocale]:
        return {entry.locale: entry for entry in self.locales}


def normalize_locale(value: object) -> str:
    locale = str(value or "").strip().replace("_", "-")
    if not _LOCALE_PATTERN.fullmatch(locale):
        raise LanguagePackError("locale must be a BCP 47 language tag such as fr-FR or ar")
    parts = locale.split("-")
    normalized = [parts[0].lower()]
    for part in parts[1:]:
        if len(part) == 2 and part.isalpha():
            normalized.append(part.upper())
        elif len(part) == 4 and part.isalpha():
            normalized.append(part.title())
        else:
            normalized.append(part)
    return "-".join(normalized)


def default_direction(locale: str) -> str:
    return "rtl" if locale.split("-", 1)[0].lower() in _RTL_LANGUAGES else "ltr"


def load_bundled_locale_manifest(directory: Path) -> BundledLocaleManifest:
    root = Path(directory).resolve()
    manifest_path = root / "manifest.json"
    try:
        raw = json.loads(manifest_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise LanguagePackError(f"cannot read bundled locale manifest: {exc}") from exc
    if not isinstance(raw, dict):
        raise LanguagePackError("bundled locale manifest must be an object")
    if raw.get("document_type") != "RLMResearchPlanner.locale-manifest":
        raise LanguagePackError("unsupported bundled locale manifest document_type")
    if int(raw.get("schema_version", 0)) != 1:
        raise LanguagePackError("unsupported bundled locale manifest schema_version")
    fallback_locale = normalize_locale(raw.get("fallback_locale") or "en-US")
    source_entries = raw.get("locales")
    if not isinstance(source_entries, list) or not source_entries:
        raise LanguagePackError("bundled locale manifest locales must be a non-empty array")
    locales: list[BundledLocale] = []
    seen: set[str] = set()
    for source in source_entries:
        if not isinstance(source, dict):
            raise LanguagePackError("bundled locale entry must be an object")
        locale = normalize_locale(source.get("locale"))
        if locale in seen:
            raise LanguagePackError(f"duplicate bundled locale: {locale}")
        name = str(source.get("name") or "").strip()
        if not name or len(name) > 100:
            raise LanguagePackError(f"{locale}: bundled locale name is invalid")
        direction = str(source.get("direction") or default_direction(locale)).strip().lower()
        if direction not in ("ltr", "rtl"):
            raise LanguagePackError(f"{locale}: direction must be ltr or rtl")
        relative = str(source.get("path") or "")
        if not relative or "\\" in relative or "/" in relative or relative in (".", ".."):
            raise LanguagePackError(f"{locale}: bundled locale path is invalid")
        path = (root / relative).resolve()
        if not path.is_relative_to(root) or not path.is_file():
            raise LanguagePackError(f"{locale}: bundled locale file is missing")
        try:
            document = json.loads(path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError) as exc:
            raise LanguagePackError(f"{locale}: bundled locale file is invalid: {exc}") from exc
        if not isinstance(document, dict) or normalize_locale(document.get("locale")) != locale:
            raise LanguagePackError(f"{locale}: bundled locale file locale does not match")
        pack = language_pack_from_dict(document, trusted=True)
        if pack.name != name or pack.direction != direction:
            raise LanguagePackError(
                f"{locale}: bundled locale metadata does not match its manifest entry"
            )
        locales.append(BundledLocale(locale, name, direction, path, pack))
        seen.add(locale)
    if fallback_locale not in seen:
        raise LanguagePackError("bundled fallback locale is not registered")
    return BundledLocaleManifest(fallback_locale, tuple(locales))


def select_preferred_locale(
    preferred_locales: Iterable[object],
    available_locales: Iterable[object],
    fallback_locale: str = "en-US",
) -> str:
    available: list[str] = []
    for value in available_locales:
        try:
            locale = normalize_locale(value)
        except LanguagePackError:
            continue
        if locale not in available:
            available.append(locale)

    fallback = normalize_locale(fallback_locale)
    if not available:
        return fallback

    preferred: list[str] = []
    for value in preferred_locales:
        try:
            locale = normalize_locale(value)
        except LanguagePackError:
            continue
        if locale not in preferred:
            preferred.append(locale)

    available_by_casefold = {locale.casefold(): locale for locale in available}
    for locale in preferred:
        exact = available_by_casefold.get(locale.casefold())
        if exact is not None:
            return exact
        language = locale.split("-", 1)[0].casefold()
        base_match = next(
            (
                candidate
                for candidate in available
                if candidate.split("-", 1)[0].casefold() == language
            ),
            None,
        )
        if base_match is not None:
            return base_match

    return available_by_casefold.get(fallback.casefold(), available[0])


def _plain_translation(
    value: object,
    *,
    section: str,
    key: str,
    trusted: bool = False,
) -> str:
    source_text = ""
    if isinstance(value, dict):
        source_text = str(value.get("source", ""))
        value = value.get("text", "")
    text = str(value or "").strip()
    if not trusted and any(character in text for character in ("<", ">")):
        raise LanguagePackError(
            f"{section}.{key}: translated text must not contain HTML markup"
        )
    if any(ord(character) < 32 and character not in "\t\n\r" for character in text):
        raise LanguagePackError(f"{section}.{key}: translated text contains a control character")
    if len(text) > 20_000:
        raise LanguagePackError(f"{section}.{key}: translated text is too long")
    if section == "messages" and text:
        try:
            translated_fields = {
                field_name
                for _literal, field_name, _format, _conversion in Formatter().parse(text)
                if field_name is not None
            }
            source_fields = {
                field_name
                for _literal, field_name, _format, _conversion in Formatter().parse(source_text)
                if field_name is not None
            }
        except ValueError as exc:
            raise LanguagePackError(
                f"{section}.{key}: translated text contains invalid braces"
            ) from exc
        if source_text and translated_fields != source_fields:
            raise LanguagePackError(
                f"{section}.{key}: translated text must keep the source placeholders"
            )
    return text


def language_pack_from_dict(raw: object, *, trusted: bool = False) -> LanguagePack:
    if not isinstance(raw, dict):
        raise LanguagePackError("language pack must be a JSON object")
    if raw.get("document_type") != LANGUAGE_PACK_DOCUMENT_TYPE:
        raise LanguagePackError("unsupported language pack document_type")
    if int(raw.get("schema_version", 0)) != LANGUAGE_PACK_SCHEMA_VERSION:
        raise LanguagePackError("unsupported language pack schema_version")
    locale = normalize_locale(raw.get("locale"))
    direction = str(raw.get("direction") or default_direction(locale)).strip().lower()
    if direction not in ("ltr", "rtl"):
        raise LanguagePackError("direction must be ltr or rtl")
    fallback_locale = normalize_locale(raw.get("fallback_locale") or "en-US")
    name = str(raw.get("name") or locale).strip()
    if not name or len(name) > 100:
        raise LanguagePackError("name is required and must be at most 100 characters")
    sections: dict[str, dict[str, str]] = {}
    for section in LANGUAGE_PACK_SECTIONS:
        source = raw.get(section, {})
        if not isinstance(source, dict):
            raise LanguagePackError(f"{section} must be an object")
        translations: dict[str, str] = {}
        for raw_key, raw_value in source.items():
            key = str(raw_key).strip()
            if not key or len(key) > 300:
                raise LanguagePackError(f"{section} contains an invalid key")
            if not trusted and section == "messages" and key in PROTECTED_MESSAGE_KEYS:
                continue
            text = _plain_translation(
                raw_value,
                section=section,
                key=key,
                trusted=trusted,
            )
            if text:
                translations[key] = text
        sections[section] = translations
    return LanguagePack(
        locale=locale,
        name=name,
        direction=direction,
        fallback_locale=fallback_locale,
        author=str(raw.get("author", "")).strip()[:200],
        license_name=str(raw.get("license", "")).strip()[:200],
        dataset_id=str(raw.get("catalog_dataset_id", "")).strip()[:200],
        effect_separator=(
            str(raw.get("effect_separator", ""))[:20]
            if "effect_separator" in raw
            else None
        ),
        sections=sections,
    )


class LanguagePackRepository:
    def __init__(self, directory: Path | None) -> None:
        self.directory = Path(directory) if directory is not None else None

    def load_all(self) -> dict[str, LanguagePack]:
        if self.directory is None or not self.directory.is_dir():
            return {}
        packs: dict[str, LanguagePack] = {}
        for path in sorted(self.directory.glob("*.json")):
            try:
                pack = language_pack_from_dict(
                    json.loads(path.read_text(encoding="utf-8"))
                )
            except (OSError, json.JSONDecodeError, LanguagePackError):
                continue
            packs[pack.locale] = pack
        return packs

    def install(self, raw: object) -> LanguagePack:
        if self.directory is None:
            raise LanguagePackError("language pack storage is unavailable")
        pack = language_pack_from_dict(raw)
        self.directory.mkdir(parents=True, exist_ok=True)
        target = self.directory / f"{pack.locale}.json"
        target.write_text(
            json.dumps(raw, ensure_ascii=False, indent=2) + "\n",
            encoding="utf-8",
        )
        return pack

    def remove(self, locale: str) -> bool:
        if self.directory is None:
            return False
        normalized = normalize_locale(locale)
        target = self.directory / f"{normalized}.json"
        if not target.is_file():
            return False
        target.unlink()
        return True


def _entry(source: str, text: str = "") -> dict[str, str]:
    return {"source": str(source), "text": str(text)}


def build_language_pack_template(
    *,
    messages: Mapping[str, str],
    master: MasterData,
    observations: Iterable[ResearchTreeObservation],
    castle_catalog: CastleCatalog,
    talent_names: Mapping[str, str] | None = None,
    talent_effects: Mapping[str, str] | None = None,
    talent_presets: Mapping[str, str] | None = None,
    talent_preset_descriptions: Mapping[str, str] | None = None,
    effect_labels: Mapping[str, str] | None = None,
    effect_values: Mapping[str, str] | None = None,
) -> dict[str, object]:
    category_sources: dict[str, str] = {}
    research_sources: dict[str, str] = {}
    effect_sources: dict[str, str] = {}
    for observation in observations:
        category_sources.setdefault(
            observation.category_id,
            observation.localized_title("en-US"),
        )
        for node in observation.nodes:
            research_sources.setdefault(node.id, node.localized_name("en-US"))
            if node.effect_label:
                effect_sources.setdefault(node.id, node.effect_label)
    for category in master.categories:
        category_sources.setdefault(
            category.id, master.localized_category(category.id, "en-US").name
        )
    for research in master.research:
        research_sources.setdefault(
            research.id, master.localized_research(research.id, "en-US").name
        )
        effect = master.localized_research(research.id, "en-US").effect_label
        if effect:
            effect_sources.setdefault(research.id, effect)
    resource_sources = {
        key.removeprefix("resource."): value
        for key, value in messages.items()
        if key.startswith("resource.")
    }
    return {
        "document_type": LANGUAGE_PACK_DOCUMENT_TYPE,
        "schema_version": LANGUAGE_PACK_SCHEMA_VERSION,
        "locale": "xx",
        "name": "New language",
        "direction": "ltr",
        "fallback_locale": "en-US",
        "author": "",
        "license": "",
        "catalog_dataset_id": master.dataset_id,
        "effect_separator": " ",
        "messages": {
            key: _entry(value)
            for key, value in sorted(messages.items())
            if key not in PROTECTED_MESSAGE_KEYS
        },
        "categories": {
            key: _entry(value) for key, value in sorted(category_sources.items())
        },
        "research": {
            key: _entry(value) for key, value in sorted(research_sources.items())
        },
        "buildings": {
            key: _entry(building.localized_name("en-US"))
            for key, building in sorted(castle_catalog.buildings.items())
        },
        "effects": {
            key: _entry(value) for key, value in sorted(effect_sources.items())
        },
        "effect_labels": {
            key: _entry(value)
            for key, value in sorted((effect_labels or {}).items())
        },
        "effect_values": {
            key: _entry(value)
            for key, value in sorted((effect_values or {}).items())
        },
        "resources": {
            key: _entry(value) for key, value in sorted(resource_sources.items())
        },
        "talents": {
            key: _entry(value)
            for key, value in sorted((talent_names or {}).items())
        },
        "talent_effects": {
            key: _entry(value)
            for key, value in sorted((talent_effects or {}).items())
        },
        "talent_presets": {
            key: _entry(value)
            for key, value in sorted((talent_presets or {}).items())
        },
        "talent_preset_descriptions": {
            key: _entry(value)
            for key, value in sorted((talent_preset_descriptions or {}).items())
        },
    }
